### ISSUE TITLE
Loading Proposals Takes Very Long

### ISSUE THREAD

**Issue by vermgit** [dbeaver/dbeaver#9377]
#### System information: 
- Windows 7
- DBeaver version 7.1.0
- Additional extensions

#### Connection specification:
- Oracle 11.2.0.4.0 - 64bit
- Driver name: Oracle
- Do you use tunnels or proxies (SSH, SOCKS, etc)? Yes

#### Describe the problem you're observing:
While typing this appears in the background processes
```
Search proposals

Seeking for SQL completion proposals: Find object by name
```
and it keeps going and going for two minutes even though I've finished typing my statement already and executed it.

After a while the proposal query seems to have received a response and the dropdown just appears in the middle of the screen even though I've moved on to some other programme. Therefore, it would make great sense to check if dbeaver still has the focus here.

In another occasion after waiting for a while dbeaver flickers, executes my statement, shows the response in the grid and then still keeps doing something that I can cancel. Knowing that "cancel" typically just stalls dbeaver I stopped pressing it a while ago. After another while the cancel-button disappears and I can finally continue to work normally.

I mean with #6310 that took about ten minutes, so it's already an improvement, but it's still far from optimal. Also note that I access a bunch of Oracle database and only one causes problems, interestingly enough our production database. Other databases pretty much take the same route in the network.

Best is if I can just turn off that proposal thing. Or let it work in the background, way way back, so it doesn't interfere with anything else. If I need a proposal I typically press Ctrl+Space to get it. Or if possible, load them after connecting to the database right away with super low priority. And increase the caching duration based on how long it took to get the information the first time.

#### Steps to reproduce, if exist:
Just write any kind of statement. At one point triggers the proposals are triggered.

#### Include any warning/errors/backtraces from the logs

```
eclipse.buildId=unknown
java.version=1.8.0_74
java.vendor=Oracle Corporation
BootLoader constants: OS=win32, ARCH=x86_64, WS=win32, NL=de_AT
Command-line arguments:  -os win32 -ws win32 -arch x86_64

org.eclipse.jface
Error
Mon Jul 27 14:38:11 CEST 2020
Problems occurred when invoking code from plug-in: "org.eclipse.jface".

java.lang.NullPointerException
	at org.eclipse.ui.internal.ide.dialogs.ProjectNaturesPage.createContents(ProjectNaturesPage.java:125)
	at org.eclipse.jface.preference.PreferencePage.createControl(PreferencePage.java:244)
	at org.eclipse.jface.preference.PreferenceDialog.createPageControl(PreferenceDialog.java:1433)
	at org.eclipse.jface.preference.PreferenceDialog$8.run(PreferenceDialog.java:1196)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.ui.internal.JFaceUtil.lambda$0(JFaceUtil.java:47)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
	at org.eclipse.jface.preference.PreferenceDialog.showPage(PreferenceDialog.java:1188)
	at org.eclipse.ui.internal.dialogs.FilteredPreferenceDialog.showPage(FilteredPreferenceDialog.java:630)
	at org.eclipse.jface.preference.PreferenceDialog$5.lambda$0(PreferenceDialog.java:660)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:72)
	at org.eclipse.jface.preference.PreferenceDialog$5.selectionChanged(PreferenceDialog.java:657)
	at org.eclipse.jface.viewers.StructuredViewer$3.run(StructuredViewer.java:823)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.ui.internal.JFaceUtil.lambda$0(JFaceUtil.java:47)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
	at org.eclipse.jface.viewers.StructuredViewer.firePostSelectionChanged(StructuredViewer.java:820)
	at org.eclipse.jface.viewers.StructuredViewer.handlePostSelect(StructuredViewer.java:1190)
	at org.eclipse.jface.viewers.StructuredViewer.lambda$0(StructuredViewer.java:1213)
	at org.eclipse.swt.events.SelectionListener$1.widgetSelected(SelectionListener.java:84)
	at org.eclipse.jface.util.OpenStrategy.firePostSelectionEvent(OpenStrategy.java:263)
	at org.eclipse.jface.util.OpenStrategy.access$5(OpenStrategy.java:258)
	at org.eclipse.jface.util.OpenStrategy$1.lambda$1(OpenStrategy.java:428)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:185)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:3897)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3527)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.ui.internal.handlers.PropertyDialogHandler.execute(PropertyDialogHandler.java:49)
	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:283)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:97)
	at sun.reflect.GeneratedMethodAccessor119.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:58)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:318)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:252)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:173)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:156)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:488)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:487)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:213)
	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.executeItem(HandledContributionItem.java:438)
	at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem.handleWidgetSelection(AbstractContributionItem.java:449)
	at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem.lambda$2(AbstractContributionItem.java:475)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4105)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1037)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:3922)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3524)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1049)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:658)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:557)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:154)
	at org.jkiss.dbeaver.ui.app.standalone.DBeaverApplication.start(DBeaverApplication.java:233)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:657)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:594)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1447)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1420)
```

**Comment by kseniiaguzeeva** [comment]
Please send me a query example to reproduce the issue.

**Comment by vermgit** [comment]
Any query does this, simple ones and complex ones.

**Comment by uslss** [comment]
@vermgit  Did you try to reset workspace? (File - Reset Workspace state)

**Comment by vermgit** [comment]
I hadn't, so I did and also updated to 7.2.0. It installed a new Oracle-driver.

The situation changed a bit: while entering the statement it takes a bit longer IMHO until that background request is being sent. And I have two "Seeking for SQL completion proposals: Find objects by name" running now as background requests that need to finish first before it actually executes my query. The query in question is:

```
SELECT *
  FROM Event e
 WHERE e.StationId IN (SELECT StationId FROM Station WHERE Sta_Name LIKE 'O%')
```

Those two proposal-requests took about a minute to complete (they finished at the same time), then they disappeared and my query was executed within like one second.

**Comment by kseniiaguzeeva** [comment]
Could you please send what Query manager is showing during the query execution (Window>Show view>Query manager).

**Comment by vermgit** [comment]
The query manager? Not the background processes?

Query manager:

![image](https://user-images.githubusercontent.com/7654880/95044921-01a60d80-06e1-11eb-90c7-a802781f0456.png)

Now that you mention it I find it interesting that it shows two commits even though I didn't commit anything.

Here're the background processes, this time they took 450 seconds before the execution of the actual query started:

![image](https://user-images.githubusercontent.com/7654880/95044824-be4b9f00-06e0-11eb-9ed5-3bf634f2d5ad.png)

**Comment by kseniiaguzeeva** [comment]
Thank you, yes, I asked about info from Query manager

**Comment by LonwoLonwo** [comment]
@vermgit Also, can you try enable this setting?
![98462637-cf4b6c80-21c6-11eb-85dc-01deec1f079b](https://user-images.githubusercontent.com/45152336/99264515-9c8e1d80-2831-11eb-8dd0-e7506394dba7.png)
It speeds up reading metadata in some cases, it can also help with Proposals

**Comment by vermgit** [comment]
Nope, didn't help. This time two queries running, the first one made the proposals appear after about two minutes, the second proposal popped up after another four minutes. Interestingly the proposals in question contained suggestions for PL/SQL-functions (To_Char and Trunc), i.e. not for table names or column names.

**Comment by vermgit** [comment]
To give an update: I'm now using version 21.0.2.202104042040 and Oracle-driver version 12.2.0.1 and it hasn't improved yet.

**Comment by LonwoLonwo** [comment]
Ok, thanks for testing.

As a workaround you can disable auto-completion settings in Preferences->Editors->SQL Editor->Code Completion tab.
You can choose "Datasource settings" to change settings only for that datasource which causes delay.

![2021-04-22 10_10_29-DBeaver Enterprise 21 0 3 - _gis_ Script-44](https://user-images.githubusercontent.com/45152336/115671322-0f587300-a353-11eb-8044-ef2f7daae6c7.png)

**Comment by kseniiaguzeeva** [comment]
@vermgit Could you tell exact queries from Query manager, which appear when the issue is happening? You should enable all query types in Query manager.
![tcO8HcVsAr](https://user-images.githubusercontent.com/51405061/119644806-fcd7ca80-be25-11eb-94ab-a9a47073747f.png)

Then double-click a query to see it completely.
![0ILNa3dFIg](https://user-images.githubusercontent.com/51405061/119644816-02351500-be26-11eb-967e-76269f8a6a61.png)

**Comment by vermgit** [comment]
Ah, OK. Well, the longest statement by far is this:

```
SELECT  DISTINCT OWNER,OBJECT_NAME,OBJECT_TYPE FROM    (SELECT OWNER,OBJECT_NAME,OBJECT_TYPE FROM ALL_OBJECTS WHERE OBJECT_TYPE IN ('TABLE','PACKAGE','PROCEDURE','FUNCTION','SYNONYM','SYNONYM') AND UPPER(OBJECT_NAME) LIKE 'STATION' UNION ALL
SELECT  O.OWNER,O.OBJECT_NAME,O.OBJECT_TYPE
FROM ALL_SYNONYMS S,ALL_OBJECTS O
WHERE O.OWNER=S.TABLE_OWNER AND O.OBJECT_NAME=S.TABLE_NAME AND S.OWNER='PUBLIC' AND UPPER(S.SYNONYM_NAME)  LIKE 'STATION')
ORDER BY OBJECT_NAME
```

as it takes 110,032 ms. The number of returned records is 1.

So I guess this means that the database maybe is missing some index or some statistics are out of order and I need to contact our database department? Or is there something else measured in the execution time that may come from DBeaver?

**Comment by p91paul** [comment]
I believe your options are:
* Check if you can query the view DBA_OBJECTS. If you can't, ask your DBA to grant your user, then enable the corresponding option in dbeaver connection preferences.
* If they deny your request, see here a new option that will be introduced with the next dbeaver version https://github.com/dbeaver/dbeaver/issues/12190#issuecomment-848550646. I expect (but did not test yet) that unchecking it will significantly improve your performance (not to the level of solution 1, but still better), however you will lose autocompletion for synonyms. 

Main point being ALL_* views (ALL_OBJECTS, ALL_SYNONYMS, ...) are horribly slow on Oracle when the number of objects in the database grows beyond trivial.

**Comment by vermgit** [comment]
I'm able to use your first option, so I'll see if that fixes it. Your last sentence would suggest that.

Thanks a lot for now, I'll get back to this if I'm still having problems.

**Comment by LonwoLonwo** [comment]
@vermgit In DBeaver 21.1.0 search for synonyms is disabled on default. Auto-complete can be faster now. Can you please test it?

![2021-06-11 13_53_38-Connection _orcl_ configuration](https://user-images.githubusercontent.com/45152336/121675779-87722800-cabc-11eb-9c47-70e48fecb760.png)

**Comment by vermgit** [comment]
After installing this version "Search metadata in synonyms" is unchecked. I guess it's the default.

My configuration also has "Use UNION for table metadata" and "Always use DBA_* views" checked (probably because of previous suggestions).

This way it's snappy, nice. However, as soon as I activate "Search metadata in synonyms" it becomes slower. So I'll keep it unchecked.

With the other two mentioned options checked I get the best results and I'm quite pleased with it. So far I haven't had any of the long requests.

AFAIC this can be closed. I'm not sure if there is anything else you want me to try so I won't close the issue.

**Comment by LonwoLonwo** [comment]
@vermgit Thanks for the feedback!

**Comment by cdichter** [comment]
Seems to be completely broken in the last two minor releases. 

Note this DB has over 400 tables, and other complexities.

**Comment by LonwoLonwo** [comment]
Hello @cdichter 

What database do you use?

**Comment by Enzodtz** [comment]
Had this issue today, changing completion engine to "legacy" solved the issue for me.
